### ISSUE TITLE
Clarity around email client implementations

### ISSUE THREAD

**Issue by bygri** [vapor/vapor#1043]
Vapor 2 shipped with a protocol for email clients. Included is an SMTP client, with configurations for Gmail, Sendgrid, and Mailgun, as well as a simple API-based client for Mailgun. There is some confusion brought in by Vapor's adoption of a simple API-based Mailgun client: does Vapor intend to supply _all_ API mail clients or just _some_ API mail clients? If the latter, then which ones? And what if community implementations overlap Vapor's own?

### Motivation

Since Vapor 1, a community package named Mail has provided API-based client implementations for Sendgrid and Mailgun, along with support for extended features supplied by these backends, such as mailing lists, embargoes, and click tracking, and sophisticated error handling. Clients for Postmark and Mandrill are forthcoming.

As maintainer of this package (now known as Airmail to avoid conflicts with Vapor 2's namespacing) I have hit conflicts with my implementation of Mailgun against Vapor's own. Do I extend Vapor's client, and sacrifice some error-handling? Do I supply a similar but not identical implementation? What if Vapor later decides to bring in Sendgrid?

### Proposed solution

I feel Vapor should supply _no_ API mail clients. Vapor should provide the `MailProtocol`, along with a basic `SMTPMailer` and an `UnimplementedMailer`. All advanced API mail clients should be provided by the community, whether by Airmail or other projects.

In short, Vapor should deprecate and remove the `Mailgun` class.

### Impact

Current projects that use Vapor's implementation of `Mailgun` would lose that functionality unless they make use of (syntactically identical) Airmail. This is a serious change I realise, but with `MailProtocol` being brought in right before the release of Vapor 2 there was little time for community discussion before 2.0 was tagged.

### Alternatives considered

Vapor could instead commit to supplying backends for all mailing backends, and my SendGrid, Postmark and Mandrill clients could be added directly into Vapor. I'd like this very much but I feel it would contribute too much 'bloat' to Vapor, especially considering how many email backends there could conceivably be.

**Comment by anthonycastelli** [comment]
@bygri and I had a chat about all of this, and I agree with his proposal. Taking Mailgun for instance there are quite a few parts of the API that would be extremely useful in many cases, i.e. mailing lists. My workaround for the Vapor 2 Mailgun implementation was this: https://gist.github.com/anthonycastelli/18bcf01d912074fd8b4b5b92b0471b1b Not a pretty solution by any means, but it does work. Having a dedicated library strictly for mail adapters would be ideal, whether it's 3rd party (Vapor Community package) or an official Vapor Mail package.

**Comment by LoganWright** [comment]
@anthonycastelli @bygri trying to evaluate possibilities here that don't involve a breaking change, and we could begin by deprecating it, and possibly moving it to its own target, we'll take some internal discussion around this, thanks for bringing it up.

**Comment by tanner0101** [comment]
I agree with this proposal. However, in terms of removing Mailgun, we can't do this until Vapor 3 since it would break API. 

For now, what @anthonycastelli shows looks like it can work as a workaround.

To fit in with other packages, I think we should create a `mailgun-provider`, `sendgrid-provider`, `postsmark-provider`, etc in the vapor-community organization.

If we have a mailgun provider ready for Vapor 2.1, we can deprecate the Mailgun class included with Vapor and instruct people to include the mailgun provider instead (which could include the extra supported functionality natively).

**Comment by anthonycastelli** [comment]
👍 Whats the planned release date for Vapor 2.1? I could probably get the malign provider created, providing 2.1 isn't due for a few weeks.

**Comment by tanner0101** [comment]
I see it as at least a few weeks away.

**Comment by bygri** [comment]
I am generally happy with this idea. I'll extract SendGrid from my mail package and make it into a Provider.

There's two other potentially useful things that were in the mail package that would now be homeless, though. One is a convenience on `EmailBody` to allow emails to be rendered from a Vapor view. The second is a pair of development-only email backends, one which logs emails to the console and another which stores emails in memory. The former, if useful, could be easily added to Vapor. Would you want one or both of the development backends added too or should there be a separate `mail-debug-provider` for those?

**Comment by bygri** [comment]
I have created [`sendgrid-provider`](https://github.com/vapor-community/sendgrid-provider) for SendGrid, and [`https://github.com/vapor-community/mail-dev-provider`](https://github.com/vapor-community/mail-dev-provider) for the development backends. The Leaf-to-email extension has been dropped.

**Comment by tanner0101** [comment]
Prioritizing this for Vapor 3: #1110. We'll create a proper mail package that includes `Email` struct (invert dependency for `SMTP`). We'll also be removing `SMTP` from being included in Vapor by default.

**Comment by MrLotU** [comment]
With Vapor 4 in Beta and no discussion since August 2017, I'm closing this issue. Feel free to re-open if there are still questions/issues unanswered.
